### PR TITLE
feat: add retry and delivery confirmation for mailgun

### PR DESCRIPTION
**Pull request by ayushnvs**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [ ] I have performed preliminary testing using the [test suite](../../apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected.
- [ ] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [ ] I have added/updated test cases to the [test suite](../../apps/api/OsmoX.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [ ] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

Added retry functionality for mailgun provider

**Related changes:**

* Added boilerplate code for common files for the retry functionality
* Added functionality to retrieve deliver status

**Screenshots:**

image

**Query request and response:**

N/A

**Documentation changes:**

N/A

**Test suite output:**

N/A

**Pending actions:**

N/A

**Additional notes:**

N/A

## Summary by CodeRabbit

- **New Features**
  - Introduced retry logic for notification delivery, including a new `retry_count` field and maximum retry attempts.
  - Added support for Mailgun delivery status tracking with success and failure states.
  
- **Enhancements**
  - Enhanced notification service to handle awaiting confirmation statuses and retry mechanisms.
  - Implemented new method in the schedule service to trigger notifications using a cron job.
  - Added configuration parameter `MAX_RETRY_COUNT` for setting maximum retry attempts.

- **Bug Fixes**
  - Improved handling of delivery status updates in notification consumers.

### CODE REVIEW

File: apps/api/src/modules/notifications/entities/notification.entity.ts [hunk 1]
```diff
@@ -81,6 +81,10 @@
   @Field()
   applicationId: number;
 
+  @Column({ name: 'retry_count', type: 'int', default: 0 })
+  @Field()
+  retryCount: number;
+
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/modules/notifications/entities/notification.entity.ts]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 2]
```diff
@@ -22,16 +24,34 @@
...
+          notification.data,
+        );
+        return this.mailgunService.sendEmail(
+          formattedNotificationData as MailgunMessageData,
+          notification.providerId,
+        );
+      });
+    } else if (notification.deliveryStatus === DeliveryStatus.AWAITING_CONFIRMATION) {
+      return super.processAwaitingConfirmationNotificationQueue(job, async () => {
+        const notificationSendResponse = notification.result.result as MessagesSendResult
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 3]
```diff
@@ -8,6 +8,8 @@
 import { MailgunMessageData } from 'mailgun.js';
 import { Notification } from 'src/modules/notifications/entities/notification.entity';
 import { NotificationConsumer } from './notification.consumer';
+import { DeliveryStatus } from 'src/common/constants/notifications';
+import { MessagesSendResult} from 'mailgun.js';
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 4]
```diff
@@ -22,16 +24,34 @@
...
+
+        const deliveryStatus = result.event;
+
+        if (deliveryStatus === 'failed' || deliveryStatus === 'rejected') {
+          return { result, deliveryStatus: DeliveryStatus.PENDING };
+        }
+
+        return { result, deliveryStatus: DeliveryStatus.SUCCESS };
+      });
+    }
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Fixed

File: apps/api/src/database/migrations/1719299457601-AddRetryCountColumn.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      type: 'int',
+      length: '3',
+      default: 0,
+    } as TableColumn);
+    // Need to handle migration for setting proper delivery_status
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.dropColumn('notify_notifications', 'retry_count');
+  }
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/database/migrations/1719299457601-AddRetryCountColumn.ts:16]: Fixed

File: apps/api/src/modules/providers/mailgun/mailgun.service.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import { Injectable, BadRequestException } from '@nestjs/common';
 import * as FormData from 'form-data';
-import Mailgun, { MailgunClientOptions, MailgunMessageData, MessagesSendResult } from 'mailgun.js';
+import Mailgun, { MailgunClientOptions, MailgunMessageData, MessagesSendResult, DomainEvent } from 'mailgun.js';
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/modules/providers/mailgun/mailgun.service.ts]: Fixed

File: apps/api/src/modules/providers/mailgun/mailgun.service.ts [hunk 7]
```diff
@@ -80,4 +80,13 @@
...
   }
+
+  async getDeliverStatus(
+    messageId: string,
+    providerId: number,
+  ): Promise<DomainEvent> {
+    await this.assignClient(providerId);
+    const response = await this.mailgunClient.events.get(this.mailgunDomain, {"message-id": messageId});
+    return response.items[0];
+  }
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/modules/providers/mailgun/mailgun.service.ts]: Fixed

File: apps/api/src/database/migrations/1719335040819-UpdateDeliveryStatusValues.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+        UPDATE notify_notifications
+        SET delivery_status = CASE
+            WHEN delivery_status = 4 THEN 3
+            WHEN delivery_status = 5 THEN 4
+            ELSE delivery_status
+        END;
+    `);
+  }
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/database/migrations/1719335040819-UpdateDeliveryStatusValues.ts:24]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 9]
```diff
@@ -22,16 +24,35 @@
...
+    if (notification.deliveryStatus === DeliveryStatus.PENDING) {
+      return super.processNotificationQueue(job, async () => {
+        const formattedNotificationData = await this.mailgunService.formatNotificationData(
+          notification.data,
+        );
+        return this.mailgunService.sendEmail(
+          formattedNotificationData as MailgunMessageData,
+          notification.providerId,
+        );
+      });
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 10]
```diff
@@ -22,16 +24,35 @@
...
+
+        const deliveryStatus = result.event;
+
+        if (deliveryStatus === 'failed' || deliveryStatus === 'rejected') {
+          return { result, deliveryStatus: DeliveryStatus.PENDING };
+        }
+
+        return { result, deliveryStatus: DeliveryStatus.SUCCESS };
+      });
+    }
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts:58]: Fixed

[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Outdated

File: apps/api/src/modules/notifications/entities/notification.entity.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/modules/notifications/entities/notification.entity.ts]: Outdated

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Outdated

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Outdated

File: apps/api/.env.example [hunk 11]
```diff
@@ -4,6 +4,9 @@ REQUEST_MAX_SIZE= # Max Size of Request, Default is 50mb
 # Node env
 NODE_ENV= # Use "development" for graphql playground to work
 
+# Notification configuration
+MAX_RETRY_COUNT= # Max retry count, default is 3
+
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/.env.example]: Outdated

File: apps/api/src/modules/notifications/queues/queue.service.ts [hunk 12]
```diff
@@ -84,9 +84,15 @@ export class QueueService {
         case `${QueueAction.SEND}-${ChannelType.SMTP}`:
           await this.smtpNotificationConsumer.processSmtpNotificationQueue(job.data.id);
           break;
+        // MAILGUN cases
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/modules/notifications/queues/queue.service.ts:87]: Error handling is already added as part of the method call.

File: apps/api/src/modules/notifications/queues/queue.service.ts [hunk 13]
```diff
@@ -84,9 +84,15 @@
...
         case `${QueueAction.SEND}-${ChannelType.MAILGUN}`:
           await this.mailgunNotificationConsumer.processMailgunNotificationQueue(job.data.id);
           break;
+        case `${QueueAction.DELIVERY_STATUS}-${ChannelType.MAILGUN}`:
+          await this.mailgunNotificationConsumer.processMailgunNotificationConfirmationQueue(
+            job.data.id,
+          );
+          break;
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/modules/notifications/queues/queue.service.ts:95]: Error handling is already added as part of the method call.

File: package-lock.json [hunk 14]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (LakshayaT) [inline package-lock.json:1]: Do we need to update this file? If not, can keep the original one only.

File: apps/api/src/modules/providers/mailgun/mailgun.service.ts [hunk 15]
```diff
@@ -80,4 +89,16 @@ export class MailgunService {
       }),
     );
   }
+
+  async getDeliverStatus(messageId: string, providerId: number): Promise<DomainEvent> {
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/mailgun/mailgun.service.ts:93]: Maybe a typo, but can you recheck if you meant to type `getDeliveryStatus`?

Can update in the method call as well in consumer.

File: package-lock.json (hunk 14 above)
Comment (ayushnvs) [reply package-lock.json]: Reverted the file

File: apps/api/src/modules/providers/mailgun/mailgun.service.ts (hunk 15 above)
Comment (ayushnvs) [reply apps/api/src/modules/providers/mailgun/mailgun.service.ts]: Fixed the typo.

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 16]
```diff
@@ -33,4 +35,31 @@
       );
     });
   }
+
+  async processMailgunNotificationConfirmationQueue(id: number): Promise<void> {
+    return super.processAwaitingConfirmationNotificationQueue(id, async () => {
+      const notification = (await this.notificationsService.getNotificationById(id))[0];
+      let notificationSendResponse = notification.result.result as MessagesSendResult;
```
[the remark this answers is not in the document]
Comment (ayushnvs) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: Fixed

File: apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 17]
```diff
@@ -33,4 +35,31 @@
...
+      const deliveryStatus = result.event;
+      notificationSendResponse.message = result.event;
+
+      if (ProviderDeliveryStatus.MAILGUN.FAILURE_STATES.includes(deliveryStatus)) {
+        return { result: notificationSendResponse, deliveryStatus: DeliveryStatus.PENDING };
+      }
+
+      if (ProviderDeliveryStatus.MAILGUN.SUCCESS_STATES.includes(deliveryStatus)) {
+        return { result: notificationSendResponse, deliveryStatus: DeliveryStatus.SUCCESS };
+      }
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts:57]: This is fine for now

**Review by LakshayaT — APPROVED** [review]

**Pull request merged**